2021/day12/Day12.java

```diff
@@ -0,0 +1,109 @@
+import java.io.File;
+import java.util.*;
+
+public class Day12 {
+    static Map<String, ArrayList<String>> neighbourCaves = new HashMap<>();
+    static Set<String> paths = new HashSet<>();
+    static Set<String> pathsPartTwo = new HashSet<>();
+    static HashMap<String, Integer> visitedCaves = new HashMap<>();
+    static HashMap<String, Integer> visitedCavesPartTwo = new HashMap<>();
+
+    static boolean isSmallCave(String cave) {
+        return cave.equals(cave.toLowerCase());
+    }
+
+    static void readInput() {
+        Scanner sc;
+        try {
+            sc = new Scanner(new File("data.txt"));
+            while (sc.hasNextLine()) {
+                String[] caves = sc.nextLine().split("-");
+                String firstCave = caves[0];
+                String secondCave = caves[1];
+
+                if (neighbourCaves.containsKey(firstCave)) {
+                    neighbourCaves.get(firstCave).add(secondCave);
+                } else {
+                    neighbourCaves.put(firstCave, new ArrayList<>(List.of(secondCave)));
+                }
+
+                if (neighbourCaves.containsKey(secondCave)) {
+                    neighbourCaves.get(secondCave).add(firstCave);
+                } else {
+                    neighbourCaves.put(secondCave, new ArrayList<>(List.of(firstCave)));
+                }
+            }
+            sc.close();
+        } catch (Exception e) {
+            System.out.println("Something went horribly wrong: " + e);
+        }
+    }
+
+    static void findPath(String currentCave, String currentPath, HashMap<String, Integer> visitedCaves) {
```

Comment:
You don't really need the `HashMap` here, do you? A `Set` or `HashSet` should suffice.

In fact, by having both a path as a `String` and the visited caves in any map/set, you duplicate a lot of information that's not necessarily needed. It should be enough to have just the current cave and the visited caves, right? The only thing `currentPath` is used for is storage.

On the other hand it would be way harder to debug if we lost the information of order, so I'm not forcing you to change this so drastically. The `HashMap` should go though.